ggml/src/ggml-cuda/unary.cu

```diff
@@ -20,7 +20,7 @@ static __global__ void step_f32(const float * x, float * dst, const int k) {
     dst[i] = x[i] > 0.0f;
 }
 
-static __global__ void gelu_f32(const float * x, float * dst, const int k) {
+static __global__ void gelu_f32(const float * __restrict__ x, float * __restrict__ dst, const int k) {
```

Comment:
I don't think not sure that restrict is correct in these functions, since they can run in-place, which would make both pointers point to the same address.

Reply:
ggml ops can only be safely run in-place if both the source and destination pointers have the same memory layout. Otherwise you would run into issues with race conditions. The use of restrict allows the compiler to reorder memory accesses during optimization under the assumption that writes to one pointer do not affect reads from another pointer. But the compiler can never optimize the code to write the result before it reads the inputs. So the use of restrict is safe with in-place ops.

My opinion is that the use of restrict should be the default. I don't feel strongly about this though and would also be fine with making the use of restrict the exception for cases where it can be demonstrated to be safe and useful.

Reply:
I disagree but I removed the use of restrict on all ops that can be used in-place, either explicitly through the `ggml.h` API or implicitly through `ggml_op_can_inplace`. On a somewhat related note, are there plans to eventually remove the `_inplace` functions in `ggml.h` and to have the logic regarding in-place operations be handled completely automatically?

Reply:
In ggml-alloc in-place is used to save memory, but in-place operations can also be useful to modify a static tensor. For example, in llama.cpp an inplace ROPE is used to perform the contexts shifts. Without inplace, this would require a normal rope followed by a cpy. In principle it would be possible to optimize away unnecessary cpy operations by making the previous operation inplace automatically, but that's not implemented, and probably not worth the complexity. So at the moment in-place operations still have a use.